File: src/51.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def generateBoard(self, n: int, queens: List[int], board: List[str]):
        retVal = []

        for i in range(n):
            board[queens[i]] = "Q"
            retVal.append("".join(board))
            board[queens[i]] = "."

        return retVal

    def backtrack(self, n: int, row: int,
                  queens: List[int], columns: set, diagonal1: set, diagonal2: set, board: List[str],
                  answer: List[List[str]]) -> None:
        if row == n:
            answer.append(self.generateBoard(n, queens, board))
            return

        for i in range(n):
            if (i in columns) or ((row - i) in diagonal1) or ((row + i) in diagonal2):
                continue

            queens[row] = i
            columns.add(i)
            diagonal1.add(row - i)
            diagonal2.add(row + i)
            self.backtrack(n, row + 1, queens, columns, diagonal1, diagonal2, board, answer)
            columns.remove(i)
            diagonal1.remove(row - i)
            diagonal2.remove(row + i)

    def solveNQueens(self, n: int) -> List[List[str]]:
        retVal = []

        queens = [-1] * n
        columns = set()
        diagonal1 = set()
        diagonal2 = set()
        board = ["."] * n
        self.backtrack(n, 0, queens, columns, diagonal1, diagonal2, board, retVal)

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for n in [4, 1]:
            # /* Example
            #  *  Input: n = 4
            #  *  Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
            #  *
            #  *  Input: n = 1
            #  *  Output: [["Q"]]
            #  */
            logging.info("Input: n = %s", n)

            retVal = pSolution.solveNQueens(n)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)

File: src/51.cpp
#include <algorithm>
#include <iostream>
#include <unordered_set>
#include <vector>

using namespace std;

class Solution {
   private:
    vector<string> generateBoard(vector<int> &queens, int n) {
        vector<string> retVal;

        for (int i = 0; i < n; i++) {
            string row = string(n, '.');
            row[queens[i]] = 'Q';
            retVal.push_back(row);
        }

        return retVal;
    }
    void backtrack(vector<vector<string>> &solutions, vector<int> &queens, int n, int row, unordered_set<int> &columns,
                   unordered_set<int> &diagonals1, unordered_set<int> &diagonals2) {
        if (row == n) {
            vector<string> board = generateBoard(queens, n);
            solutions.push_back(board);
            return;
        }

        for (int i = 0; i < n; i++) {
            if (columns.find(i) != columns.end()) {
                continue;
            }

            int diagonal1 = row - i;
            if (diagonals1.find(diagonal1) != diagonals1.end()) {
                continue;
            }

            int diagonal2 = row + i;
            if (diagonals2.find(diagonal2) != diagonals2.end()) {
                continue;
            }

            queens[row] = i;
            columns.insert(i);
            diagonals1.insert(diagonal1);
            diagonals2.insert(diagonal2);
            backtrack(solutions, queens, n, row + 1, columns, diagonals1, diagonals2);
            queens[row] = -1;
            columns.erase(i);
            diagonals1.erase(diagonal1);
            diagonals2.erase(diagonal2);
        }
    }

   public:
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> retVal;

        auto queens = vector<int>(n, -1);
        auto columns = unordered_set<int>();
        auto diagonals1 = unordered_set<int>();
        auto diagonals2 = unordered_set<int>();
        backtrack(retVal, queens, n, 0, columns, diagonals1, diagonals2);

        return retVal;
    }
};

int main(int argc, char **argv) {
    struct subject {
        int n;
    };
    vector<subject> testData{{4}, {1}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input: n = 4
     *  Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
     *
     *  Input: n = 1
     *  Output: [["Q"]]
     */

    Solution cSolution;
    vector<vector<string>> answer;
    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input: n = " << testData[i].n << "\n";

        answer = cSolution.solveNQueens(testData[i].n);
        cout << "Output: [";
        for (long unsigned int j = 0; j < answer.size(); ++j) {
            cout << (j == 0 ? "" : ",") << "[";
            for (long unsigned int k = 0; k < answer[j].size(); ++k) {
                cout << (k == 0 ? "" : ",") << "\"" << answer[j][k] << "\"";
            }
            cout << "]";
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}
